src/v/pandaproxy/schema_registry/avro.cc

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/schema_registry/avro.h"
+
+#include "pandaproxy/schema_registry/error.h"
+
+#include <avro/Compiler.hh>
+#include <avro/Exception.hh>
+#include <avro/GenericDatum.hh>
+#include <avro/ValidSchema.hh>
+
+namespace pandaproxy::schema_registry {
+
+result<avro_schema_definition>
+make_avro_schema_definition(std::string_view sv) {
+    try {
+        return avro_schema_definition{avro::compileJsonSchemaFromMemory(
+          reinterpret_cast<const uint8_t*>(sv.data()), sv.length())};
+    } catch (const avro::Exception& e) {
+        return error_code::schema_invalid;
+    }
+}
+
+bool check_compatible(
+  const avro::NodePtr& reader, const avro::NodePtr& writer) {
```

Comment:
is this a `const foo*&reader` ? why not pass the const ptr

Reply:
It's a `std::shared_ptr<avro::Node>`

Reply:
so why not copy the pointer. const ptr. why ptr-ref ?

Reply:
None of this is futurized, so there's no lifetime issue. Making a copy would hit the atomic count on every recursion; taking by reference avoids that.